File: src/factories/internal/metadata/iterate_metadata_alias.ts
import ts from "typescript";

import { MetadataAlias } from "../../../schemas/metadata/MetadataAlias";
import { MetadataAliasType } from "../../../schemas/metadata/MetadataAliasType";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { emplace_metadata_alias } from "./emplace_metadata_alias";

export const iterate_metadata_alias = (
  props: IMetadataIteratorProps,
): boolean => {
  if (props.options.absorb !== false || props.type.aliasSymbol === undefined)
    return false;

  const node: ts.Declaration | undefined =
    props.type.aliasSymbol.declarations?.[0];
  if (node === undefined) return false;

  // CONSTRUCT DEFINITION
  const type: MetadataAliasType = emplace_metadata_alias(props);
  ArrayUtil.take(
    props.metadata.aliases,
    (elem) => elem.type.name === type.name,
    () =>
      MetadataAlias.create({
        type,
        tags: [],
      }),
  );
  return true;
};

File: packages/core/src/factories/internal/metadata/iterate_metadata_alias.go
package metadata

func IterateMetadataAlias(props IMetadataIteratorProps) bool {
	if props.Options.Absorb != nil && *props.Options.Absorb {
		return false
	}
	if props.Type == nil || props.Type.AliasSymbol() == nil {
		return false
	}
	alias := EmplaceMetadataAlias(props)
	takeAlias(&props.Metadata.Aliases, MetadataAlias{Type: alias})
	return true
}
